magmi/plugins/extra/itemprocessors/imageprocessor/imageitattributeemprocessor.php

```diff
@@ -224,29 +224,38 @@ public function handleVarcharAttribute($pid, &$item, $storeid, $attrcode, $attrd
     }
 
     /**
-     * imageInGallery
+     * Creates or updates image and returns its id.
      *
      * @param int $pid
      *            : product id to test image existence in gallery
      * @param string $imgname
      *            : image file name (relative to /products/media in magento dir)
+     * @param int $refid
+     *            : A reference to an attribute type, typically of image,
+     *            small_image or thumbnail
      * @return bool : if image is already present in gallery for a given product id
      */
     public function getImageId($pid, $attid, $imgname, $refid = null, $store_id = 0)
     {
         $t = $this->tablename('catalog_product_entity_media_gallery');
 
         $sql = "SELECT $t.value_id FROM $t ";
+
+        // Try finding matching refid
         if ($refid != null) {
             $vc = $this->tablename('catalog_product_entity_varchar');
-            $sql .= " JOIN $vc ON $t.entity_id=$vc.entity_id AND $t.value=$vc.value AND $vc.attribute_id=?
+            $full_sql = $sql . " JOIN $vc ON $t.entity_id=$vc.entity_id AND $t.value=$vc.value AND $vc.attribute_id=?
 					WHERE $t.entity_id=? AND $vc.store_id=?";
-            $imgid = $this->selectone($sql, array($refid, $pid, $store_id), 'value_id');
-        } else {
-            $sql .= " WHERE value=? AND entity_id=? AND attribute_id=?";
-            $imgid = $this->selectone($sql, array($imgname, $pid, $attid), 'value_id');
+            $imgid = $this->selectone($full_sql, array($refid, $pid, $store_id), 'value_id');
+        }
+
+        // If not found just match on type and value
+        if ($imgid == null) {
+            $full_sql = $sql . " WHERE value=? AND entity_id=? AND attribute_id=?";
+            $imgid = $this->selectone($full_sql, array($imgname, $pid, $attid), 'value_id');
```

Comment:
You are not intended to have multiple set of media gallery images for each store. 

Although you can create that using Magmi it is kind of on borrowed time since when you go and save a product in the product view for which you didn't upload the image Magento will simply insert a new row to `catalog_product_entity_media_gallery_value`. 

You don't even have to be on the Images tab. It will happen no matter what. The same if you remove an image from a store view it will be removed from all store views. A product has a media gallery with same images

What you CAN do is set 'exclude' differently for each store view so that an image is excluded from some and not for others. (not now of course since the code pre this PR would create duplicates, that's later removed by Magento). [See this article](https://www.mag-manager.com/useful-articles/how-to/add-different-images-to-differed-magento-store-views/)

So coming to think of it you might not even need to do the join with `catalog_product_entity_varchar` first. (I'll try that out).

ps: when I was talking about duplicates and uniqueness it was not with regards to system wide images but only on the same product, which is where the problem is.